README.md

# easylift

<!-- badges: start -->

[![GitHub
issues](https://img.shields.io/github/issues/nahid18/easylift)](https://github.com/nahid18/easylift/issues)
[![GitHub
pulls](https://img.shields.io/github/issues-pr/nahid18/easylift)](https://github.com/nahid18/easylift/pulls)
<!-- badges: end -->

The goal of `easylift` is to perform genomic liftover given `GRanges`
and `chain` file.

## Installation

``` r
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("easylift")
```

## Documentation

To view documentation:

``` r
browseVignettes("easylift")
```

## Import

Import the libraries

``` r
library("easylift")
```

Call `easylift` with `GRanges` object, target genome and the chain file.

``` r
gr <- GRanges(
  seqname = Rle(
    c("chr1", "chr2"), 
    c(100000, 100000)
  ),
  ranges = IRanges(
    start = 1, 
    end = 200000
  )
)
# Here, "hg19" is the source genome
genome(gr) <- "hg19"
chain <- "hg19ToHg38.over.chain.gz"

# Here, "hg38" is the target genome
easylift(gr, "hg38", chain)
```

### BiocFileCache

To use `BiocFileCache` for the chain file, add it to the cache:

``` r
chain_file <- "/path/to/your/hg19ToHg38.over.chain.gz"
bfc <- BiocFileCache()

# Add chain file to cache if already not available
if (nrow(bfcquery(bfc, basename(chain_file))) == 0)
    bfcadd(bfc, chain_file)
```

Then, use it in `easylift`:

``` r
easylift(gr, "hg38") 
# or
gr |> easylift("hg38") 
```

## Citation

To cite package `easylift` in publications use:

Al Nahid A, Pagès H, Love M (2023). *easylift: An R package to perform
genomic liftover*. R package version 1.1.0,
<https://github.com/nahid18/easylift>.

A BibTeX entry for LaTeX users is

      @Manual{,
        title = {easylift: An R package to perform genomic liftover},
        author = {Abdullah Al Nahid, Hervé Pagès, Michael Love},
        year = {2023},
        note = {R package version 1.1.0},
        url = {https://github.com/nahid18/easylift},
      }

Please note that the `easylift` was only made possible thanks to many
other R and bioinformatics software authors, which are cited either in
the vignettes and/or the paper(s) describing this package.

## Code of Conduct

Please note that the `easylift` project is released with a [Contributor
Code of Conduct](http://bioconductor.org/about/code-of-conduct/). By
contributing to this project, you agree to abide by its terms.


NEWS.md
# easylift 1.0.0

* Bioconductor release 3.18


# easylift 0.99.9

* Bug fixes and further improvements to the code.


# easylift 0.99.1

* Updated the code and documentation to reflect the changes suggested by the Bioconductor team.


# easylift 0.99.0

* Submitted to Bioconductor


tests/testthat.R
# This file is part of the standard setup for testthat.
# It is recommended that you do not modify it.
#
# Where should you do additional test configuration?
# Learn more about the roles of various files in:
# * https://r-pkgs.org/testing-design.html#sec-tests-files-overview
# * https://testthat.r-lib.org/articles/special-files.html

library(testthat)
library(easylift)

test_check("easylift")


tests/testthat/test-easylift.R
library(GenomicRanges)
library(BiocFileCache)
library(IRanges)

test_that("easylift function tests with valid chain files", {
  # Test with a valid chain gzipped file
  chain_path <- system.file("extdata", "hg19ToHg38.over.chain", package = "easylift")
  gr1 <- GenomicRanges::GRanges(
    seqnames = "chr1",
    ranges = IRanges::IRanges(start = 100, end = 200),
    strand = "+"
  )
  genome(gr1) <- "hg19"
  expect_type(easylift(x = gr1, to = "hg38", chain = chain_path), "S4")
  expect_no_error(easylift(x = gr1, to = "hg38", chain = chain_path))
})

test_that("easylift function tests with error cases", {
  # Test with an empty GRanges object
  gr3 <- GenomicRanges::GRanges()
  expect_error(easylift(x = gr3, to = "hg38", chain = chain_path))

  # Test with missing genome information
  gr4 <- GenomicRanges::GRanges(
    seqnames = "chr4",
    ranges = IRanges::IRanges(start = 400, end = 500),
    strand = "+"
  )
  expect_error(easylift(x = gr4, to = "hg38", chain = chain_path))

  # Test with missing genome information
  gr5 <- GenomicRanges::GRanges(
    seqnames = "chr4",
    ranges = IRanges::IRanges(start = 400, end = 500),
    strand = "+"
  )
  genome(gr5) <- "hg19"
  expect_error(easylift(x = gr5, to = NULL, chain = chain_path))

  # Test with missing genome information
  gr6 <- GenomicRanges::GRanges(
    seqnames = "chr4",
    ranges = IRanges::IRanges(start = 400, end = 500),
    strand = "+"
  )
  genome(gr6) <- "hg19"
  expect_error(easylift(x = gr6, to = NULL))
})

test_that("easylift succeeds with BiocFileCache", {
  # Create a BiocFileCache instance (to a temp location for this test script)
  path <- tempfile()
  bfc <- BiocFileCache(path, ask = FALSE)

  chain_file <- system.file("extdata", "hg19ToHg38.over.chain", package = "easylift")

  # Add the test chain file to the cache
  bfcadd(bfc, chain_file)

  # Query the cache for the chain file
  q <- bfcquery(bfc, chain_file)

  # Test if the query result has rows (file exists in the cache)
  expect_true(nrow(q) > 0, "Chain file should exist in cache.")

  # Create a test GRanges object
  gr <- GenomicRanges::GRanges(seqname = Rle(c("chr1", "chr2"), c(100000, 100000)),
                               ranges = IRanges::IRanges(start = 1, end = 200000))

  genome(gr) <- "hg19"

  # Should pass since bfc is provided properly to easylift
  expect_no_error(easylift(gr, "hg38", NULL, bfc))
  expect_no_error(gr |> easylift("hg38", NULL, bfc))
  expect_no_error(easylift(x = gr, to = "hg38", bfc = bfc))

  # Test error when provided bfc is not a BiocFileCache instance
  expect_error(easylift(gr, "hg38", bfc))
  expect_error(gr |> easylift("hg38", bfc))

  # Should error because the chain file was added to tempfile() but not in the
  # default cache location
  expect_error(easylift(gr, "hg38"))
  expect_error(gr |> easylift("hg38"))

  # Test error when provided bfc is not a BiocFileCache instance
  expect_error(easylift(x = gr, to = "hg38", bfc = "bfc"))

  # Test error when chain file not found
  expect_error(easylift(x = gr, to = "hg100"))

  # Test error when invalid chain file provided
  expect_error(easylift(x = gr, to = "hg100", chain="/random/path/not-found.chain"))
})


R/easylift.R
#' Lift genomic coordinates from one genome assembly to another.
#'
#' This function takes a GRanges object with genomic coordinates in one genome
#' assembly and lifts them to target genome assembly using a chain file.
#'
#' @param x A GRanges object with genomic coordinates in the original assembly.
#' @param to The target genome assembly (e.g., "hg38").
#' @param chain The path to the chain file containing the liftover mapping.
#' Can be provided in gzipped or non-gzipped format. If omitted, the function
#' will look in the default BiocFileCache for a properly named chain file.
#' @param bfc A BiocFileCache object (optional), if not provided (most typically)
#' the default location will be used.
#'
#' @return A GRanges object with lifted genomic coordinates.
#'
#' @examples
#' # Lift over the coordinates of the first 10 genes in the hg19 assembly
#' # to the hg38 assembly
#' library(easylift)
#' gr <- GRanges(
#'   seqname = Rle(c("chr1", "chr2"), c(100000, 100000)),
#'   ranges = IRanges(start = 1, end = 200000)
#' )
#' # Here, "hg19" is the source genome
#' genome(gr) <- "hg19"
#'
#' # Here, we use the `system.file()` function because the chain file is in the
#' # package (however if you need to point to any other file on your machine,
#' # just do 'chain <- "path/to/your/hg19ToHg38.over.chain.gz"'):
#' chain <- system.file("extdata", "hg19ToHg38.over.chain.gz", package = "easylift")
#'
#' # Here, "hg38" is the target genome
#' easylift(gr, "hg38", chain)
#'
#' \donttest{
#' # To use `BiocFileCache` for the chain file, add it to the cache as follows:
#' chain_file <- "/path/to/your/hg19ToHg38.over.chain.gz"
#' bfc <- BiocFileCache()
#'
#' # Add chain file to cache if already not available
#' if (nrow(bfcquery(bfc, basename(chain_file))) == 0)
#'    bfcadd(bfc, chain_file)
#'
#' # Then, use it in `easylift` like this:
#'
#' easylift(gr, "hg38")
#' # or
#' gr |> easylift("hg38")
#' }
#' @import GenomicRanges
#' @import GenomeInfoDb
#' @import rtracklayer
#' @import R.utils
#' @import tools
#' @import BiocFileCache
#' @import methods
#' @seealso
#' \code{\link[rtracklayer]{liftOver}} function from the \code{rtracklayer} package,
#' which is the basis for \code{easylift}.
#' @export
easylift <- function(x, to, chain, bfc) {
  args <- .perform_sanity_check(x, to, chain, bfc)
  x <- args$x
  to <- args$to
  chain <- args$chain
  bfc <- args$bfc
  unique_genomes <- args$unique_genomes

  # Convert the input GRanges to the "UCSC" seqlevels style if not already
  if (GenomeInfoDb::seqlevelsStyle(x) != "UCSC") {
    GenomeInfoDb::seqlevelsStyle(x) <- "UCSC"
  }

  if (.is_faulty(chain)) {
    chain <- .get_chain_from_BiocFileCache(unique_genomes, to, bfc)
  }

  # Check if the chain file is gzipped and unzip if needed
  if (tools::file_ext(chain) == "gz") {
    tmp_dir <- tempdir()
    unzipped_chain <- file.path(tmp_dir, "tmp.chain")
    R.utils::gunzip(chain,
                    destname = unzipped_chain,
                    overwrite = TRUE,
                    remove = FALSE)
    chain <- unzipped_chain
  }

  # Load the chain file
  ch <- rtracklayer::import.chain(chain)

  # Check if the 'to' genome is valid and available
  to_seqinfo <-
    GenomeInfoDb::getChromInfoFromUCSC(to,
                                       assembled.molecules.only = TRUE,
                                       as.Seqinfo = TRUE)
  if (is.null(to_seqinfo)) {
    stop("The genome", to, "is not available or recognized.")
  }

  # LiftOver the genomic coordinates
  cur <- unlist(rtracklayer::liftOver(x, ch))

  GenomeInfoDb::seqlevels(cur) <-
    GenomeInfoDb::seqlevels(to_seqinfo)
  GenomeInfoDb::seqinfo(cur) <- to_seqinfo

  return(cur)
}

### 'from' and 'to' are single strings of UCSC genome names, and 'bfc' is a BiocFileCache object
.get_chain_from_BiocFileCache <- function(from, to, bfc) {
  is_bfc_faulty <- .is_faulty(bfc)
  if (is_bfc_faulty) {
    bfc <- BiocFileCache()
  }
  capTo <-
    paste0(toupper(substr(to, 1, 1)), substr(to, 2, nchar(to)))
  trychainfile <- paste0(from, "To", capTo, ".over.chain")
  q <- bfcquery(bfc, trychainfile)
  if (nrow(q) == 0) {
    bfc_str <- if (is_bfc_faulty)
      "default"
    else
      "provided"
    stop(
      "Chain file not specified and filename with ",
      trychainfile,
      " pattern not found in BiocFileCache ",
      bfc_str,
      " location."
    )
  }
  chain <- bfc[[q$rid[1]]]
  return(chain)
}

# chain is expected to be string path of a chain file
.is_faulty <- function(y) {
  return(missing(y) || is.null(y))
}

.is_available <- function(y) {
  return(!missing(y) && !is.null(y))
}

.perform_sanity_check <- function(x, to, chain, bfc) {
  if (missing(chain)) {
    chain <- NULL
  }
  if (missing(bfc)) {
    bfc <- NULL
  }
  if (.is_faulty(x)) {
    stop("Please provide a 'GRanges' object 'x' with genomic coordinates.")
  }
  if (!is(x, "GRanges")) {
    stop("Please provide a valid 'GRanges' object 'x'.")
  }
  if (.is_faulty(to)) {
    stop("Please provide the target genome assembly 'to'.")
  }
  if (!is.character(to)) {
    stop("Please provide the target genome assembly 'to' as a string.")
  }
  if (.is_available(chain)) {
    if (!is.character(chain)) {
      stop("Please provide the chain file path 'chain' as a string.")
    }
  }
  if (.is_available(bfc)) {
    if (!is(bfc, "BiocFileCache")) {
      stop("Please provide a 'BiocFileCache' object 'bfc'.")
    }
  }
  if (anyNA(GenomeInfoDb::genome(x))) {
    stop("The genome information is missing. Please set genome(x) before using easylift.")
  }
  unique_genomes <- unique(GenomeInfoDb::genome(x))
  if (length(unique_genomes) > 1) {
    stop(
      "The 'GRanges' object 'x' contains genomic coordinates from multiple genomes. ",
      "Please provide 'x' with coordinates from a single genome assembly."
    )
  }
  return(list(
    x = x,
    to = to,
    chain = chain,
    bfc = bfc,
    unique_genomes = unique_genomes
  ))
}
